test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace EdgeHubRestartTester
+{
+    using System;
+    using System.Net;
+    using System.Net.Sockets;
+    using System.Text;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Client.Exceptions;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    class DirectMethodEdgeHubConnector : IEdgeHubConnector
+    {
+        long directMethodCount = 0;
+        ModuleClient dmModuleClient;
+        Guid batchId;
+        DateTime runExpirationTime;
+        CancellationToken cancellationToken;
+        DateTime edgeHubRestartedTime;
+        uint restartSequenceNumber;
+        ILogger logger;
+
+        public DirectMethodEdgeHubConnector(
+            ModuleClient dmModuleClient,
+            Guid batchId,
+            DateTime runExpirationTime,
+            DateTime edgeHubRestartedTime,
+            uint restartSequenceNumber,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            this.dmModuleClient = Preconditions.CheckNotNull(dmModuleClient, nameof(dmModuleClient));
+            this.batchId = batchId;
+            this.runExpirationTime = runExpirationTime;
+            this.cancellationToken = Preconditions.CheckNotNull(cancellationToken, nameof(cancellationToken));
+            this.edgeHubRestartedTime = edgeHubRestartedTime;
+            this.restartSequenceNumber = restartSequenceNumber;
+            this.logger = Preconditions.CheckNotNull(logger, nameof(logger));
+        }
+
+        public async Task StartAsync()
+        {
+            (DateTime dmCompletedTime, HttpStatusCode dmStatusCode) = await this.SendDirectMethodAsync(
+                Settings.Current.DeviceId,
+                Settings.Current.DirectMethodTargetModuleId,
+                this.dmModuleClient,
+                Settings.Current.DirectMethodName,
+                this.runExpirationTime,
+                this.cancellationToken,
+                this.logger).ConfigureAwait(false);
+
+            TestResultBase dmTestResult = new EdgeHubRestartDirectMethodResult(
+                Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString(),
+                DateTime.UtcNow,
+                Settings.Current.TrackingId,
+                this.batchId,
+                Interlocked.Read(ref this.directMethodCount).ToString(),
+                this.edgeHubRestartedTime,
+                dmCompletedTime,
+                dmStatusCode,
+                this.restartSequenceNumber);
+
+            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
```

Comment:
I thought we discussed returning the TestResultBase to the caller. This is fine, but any reason why you didn't go that route?

Reply:
Since the module send Direct Method and Message are independent of each other, I prefer to have each DirectMethod and Message Task be responsible for its own result reporting without depending if their counterpart fail/success/die.

Reply:
Its fine for now. But a cleaner way would be to limit each class to do one thing.

Reply:
Deferring the comment to : https://msazure.visualstudio.com/One/_workitems/edit/6186177